// @microsoft/teams.apps/dist/middleware/auth/service-token-validator.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ServiceTokenValidator = void 0;
const teams_api_1 = require("@microsoft/teams.api");
const jwt_validator_1 = require("./jwt-validator");
/**
 * Derives the JWKS keys URI from an OpenID metadata URL.
 * e.g. "https://login.botframework.com/v1/.well-known/openidconfiguration"
 *   -> "https://login.botframework.com/v1/.well-known/keys"
 */
function openIdMetadataToKeysUri(openIdMetadataUrl) {
    return openIdMetadataUrl.replace(/\/openidconfiguration$/, '/keys');
}
/**
 * Service token validator for Bot Framework /api/messages requests
 * Validates Bot Framework service tokens
 */
class ServiceTokenValidator {
    jwtValidator;
    credentials;
    constructor(appId, tenantId, serviceUrl, logger, cloud) {
        const env = cloud ?? teams_api_1.PUBLIC;
        this.jwtValidator = new jwt_validator_1.JwtValidator({
            clientId: appId,
            tenantId,
            loginEndpoint: env.loginEndpoint,
            validateIssuer: { allowedIssuer: env.tokenIssuer },
            validateServiceUrl: serviceUrl ? { expectedServiceUrl: serviceUrl } : undefined,
            jwksUriOptions: {
                type: 'uri',
                uri: openIdMetadataToKeysUri(env.openIdMetadataUrl),
            },
        }, logger);
        this.credentials = { clientId: appId, tenantId };
    }
    async check(authHeader, body) {
        // Extract token from "Bearer <token>" format
        const token = authHeader.startsWith('Bearer ')
            ? authHeader.substring(7)
            : authHeader;
        // Validate token using JWT validator
        const payload = await this.jwtValidator.validateAccessToken(token, {
            validateServiceUrl: body.serviceUrl ? { expectedServiceUrl: body.serviceUrl } : undefined
        });
        if (!payload) {
            throw new Error('Invalid token');
        }
        const serviceUrl = body.serviceUrl || payload.serviceurl || '';
        // Convert JWT payload to IToken
        return {
            appId: payload.appid || this.credentials?.clientId || '',
            from: 'azure',
            fromId: payload.sub || '',
            serviceUrl,
            isExpired: () => false, // Already validated by JWT validator
        };
    }
}
exports.ServiceTokenValidator = ServiceTokenValidator;
