// @microsoft/teams.api/dist/clients/index.mjs
import { Client as Client$1 } from '@microsoft/teams.common';
import { mergeApiClientSettings } from './api-client-settings.mjs';
export * from './api-client-settings.mjs';
import { BotClient } from './bot/index.mjs';
export * from './bot/index.mjs';
import { ConversationClient } from './conversation/index.mjs';
export * from './conversation/index.mjs';
import { MeetingClient } from './meeting.mjs';
export * from './meeting.mjs';
import { ReactionClient } from './reaction/index.mjs';
export * from './reaction/index.mjs';
import { TeamClient } from './team.mjs';
export * from './team.mjs';
import { UserClient } from './user/index.mjs';
export * from './user/index.mjs';

class Client {
  serviceUrl;
  bots;
  users;
  conversations;
  teams;
  meetings;
  reactions;
  get http() {
    return this._http;
  }
  set http(v) {
    this.bots.http = v;
    this.conversations.http = v;
    this.users.http = v;
    this.teams.http = v;
    this.meetings.http = v;
    this.reactions.http = v;
    this._http = v;
  }
  _http;
  _apiClientSettings;
  constructor(serviceUrl, options, apiClientSettings, cloud) {
    this.serviceUrl = serviceUrl;
    if (!options) {
      this._http = new Client$1();
    } else if ("request" in options) {
      this._http = options;
    } else {
      this._http = new Client$1({
        ...options,
        headers: {
          ...options?.headers,
          "Content-Type": "application/json"
        }
      });
    }
    this._apiClientSettings = mergeApiClientSettings(apiClientSettings, cloud);
    this.bots = new BotClient(this.http, this._apiClientSettings);
    this.users = new UserClient(this.http, this._apiClientSettings);
    this.conversations = new ConversationClient(serviceUrl, this.http, this._apiClientSettings);
    this.teams = new TeamClient(serviceUrl, this.http, this._apiClientSettings);
    this.meetings = new MeetingClient(serviceUrl, this.http, this._apiClientSettings);
    this.reactions = new ReactionClient(serviceUrl, this.http, this._apiClientSettings);
  }
}

export { Client };
