// @microsoft/teams.api/dist/clients/index.js
'use strict';

var teams_common = require('@microsoft/teams.common');
var apiClientSettings = require('./api-client-settings');
var bot = require('./bot');
var conversation = require('./conversation');
var meeting = require('./meeting');
var reaction = require('./reaction');
var team = require('./team');
var user = require('./user');

class Client {
  serviceUrl;
  bots;
  users;
  conversations;
  teams;
  meetings;
  reactions;
  get http() {
    return this._http;
  }
  set http(v) {
    this.bots.http = v;
    this.conversations.http = v;
    this.users.http = v;
    this.teams.http = v;
    this.meetings.http = v;
    this.reactions.http = v;
    this._http = v;
  }
  _http;
  _apiClientSettings;
  constructor(serviceUrl, options, apiClientSettings$1, cloud) {
    this.serviceUrl = serviceUrl;
    if (!options) {
      this._http = new teams_common.Client();
    } else if ("request" in options) {
      this._http = options;
    } else {
      this._http = new teams_common.Client({
        ...options,
        headers: {
          ...options?.headers,
          "Content-Type": "application/json"
        }
      });
    }
    this._apiClientSettings = apiClientSettings.mergeApiClientSettings(apiClientSettings$1, cloud);
    this.bots = new bot.BotClient(this.http, this._apiClientSettings);
    this.users = new user.UserClient(this.http, this._apiClientSettings);
    this.conversations = new conversation.ConversationClient(serviceUrl, this.http, this._apiClientSettings);
    this.teams = new team.TeamClient(serviceUrl, this.http, this._apiClientSettings);
    this.meetings = new meeting.MeetingClient(serviceUrl, this.http, this._apiClientSettings);
    this.reactions = new reaction.ReactionClient(serviceUrl, this.http, this._apiClientSettings);
  }
}

exports.Client = Client;
Object.keys(apiClientSettings).forEach(function (k) {
  if (k !== 'default' && !Object.prototype.hasOwnProperty.call(exports, k)) Object.defineProperty(exports, k, {
    enumerable: true,
    get: function () { return apiClientSettings[k]; }
  });
});
Object.keys(bot).forEach(function (k) {
  if (k !== 'default' && !Object.prototype.hasOwnProperty.call(exports, k)) Object.defineProperty(exports, k, {
    enumerable: true,
    get: function () { return bot[k]; }
  });
});
Object.keys(conversation).forEach(function (k) {
  if (k !== 'default' && !Object.prototype.hasOwnProperty.call(exports, k)) Object.defineProperty(exports, k, {
    enumerable: true,
    get: function () { return conversation[k]; }
  });
});
Object.keys(meeting).forEach(function (k) {
  if (k !== 'default' && !Object.prototype.hasOwnProperty.call(exports, k)) Object.defineProperty(exports, k, {
    enumerable: true,
    get: function () { return meeting[k]; }
  });
});
Object.keys(reaction).forEach(function (k) {
  if (k !== 'default' && !Object.prototype.hasOwnProperty.call(exports, k)) Object.defineProperty(exports, k, {
    enumerable: true,
    get: function () { return reaction[k]; }
  });
});
Object.keys(team).forEach(function (k) {
  if (k !== 'default' && !Object.prototype.hasOwnProperty.call(exports, k)) Object.defineProperty(exports, k, {
    enumerable: true,
    get: function () { return team[k]; }
  });
});
Object.keys(user).forEach(function (k) {
  if (k !== 'default' && !Object.prototype.hasOwnProperty.call(exports, k)) Object.defineProperty(exports, k, {
    enumerable: true,
    get: function () { return user[k]; }
  });
});
