// @microsoft/teams.api/dist/auth/cloud-environment.js
'use strict';

const PUBLIC = Object.freeze({
  loginEndpoint: "https://login.microsoftonline.com",
  loginTenant: "botframework.com",
  botScope: "https://api.botframework.com/.default",
  tokenServiceUrl: "https://token.botframework.com",
  openIdMetadataUrl: "https://login.botframework.com/v1/.well-known/openidconfiguration",
  tokenIssuer: "https://api.botframework.com",
  graphScope: "https://graph.microsoft.com/.default"
});
const US_GOV = Object.freeze({
  loginEndpoint: "https://login.microsoftonline.us",
  loginTenant: "MicrosoftServices.onmicrosoft.us",
  botScope: "https://api.botframework.us/.default",
  tokenServiceUrl: "https://tokengcch.botframework.azure.us",
  openIdMetadataUrl: "https://login.botframework.azure.us/v1/.well-known/openidconfiguration",
  tokenIssuer: "https://api.botframework.us",
  graphScope: "https://graph.microsoft.us/.default"
});
const US_GOV_DOD = Object.freeze({
  loginEndpoint: "https://login.microsoftonline.us",
  loginTenant: "MicrosoftServices.onmicrosoft.us",
  botScope: "https://api.botframework.us/.default",
  tokenServiceUrl: "https://apiDoD.botframework.azure.us",
  openIdMetadataUrl: "https://login.botframework.azure.us/v1/.well-known/openidconfiguration",
  tokenIssuer: "https://api.botframework.us",
  graphScope: "https://dod-graph.microsoft.us/.default"
});
const CHINA = Object.freeze({
  loginEndpoint: "https://login.partner.microsoftonline.cn",
  loginTenant: "microsoftservices.partner.onmschina.cn",
  botScope: "https://api.botframework.azure.cn/.default",
  tokenServiceUrl: "https://token.botframework.azure.cn",
  openIdMetadataUrl: "https://login.botframework.azure.cn/v1/.well-known/openidconfiguration",
  tokenIssuer: "https://api.botframework.azure.cn",
  graphScope: "https://microsoftgraph.chinacloudapi.cn/.default"
});
function withOverrides(base, overrides) {
  return Object.freeze({
    loginEndpoint: overrides.loginEndpoint ?? base.loginEndpoint,
    loginTenant: overrides.loginTenant ?? base.loginTenant,
    botScope: overrides.botScope ?? base.botScope,
    tokenServiceUrl: overrides.tokenServiceUrl ?? base.tokenServiceUrl,
    openIdMetadataUrl: overrides.openIdMetadataUrl ?? base.openIdMetadataUrl,
    tokenIssuer: overrides.tokenIssuer ?? base.tokenIssuer,
    graphScope: overrides.graphScope ?? base.graphScope
  });
}
const CLOUD_ENVIRONMENTS = {
  public: PUBLIC,
  usgov: US_GOV,
  usgovdod: US_GOV_DOD,
  china: CHINA
};
function cloudFromName(name) {
  const env = CLOUD_ENVIRONMENTS[name.toLowerCase()];
  if (!env) {
    throw new Error(
      `Unknown cloud environment: '${name}'. Valid values are: Public, USGov, USGovDoD, China.`
    );
  }
  return env;
}

exports.CHINA = CHINA;
exports.PUBLIC = PUBLIC;
exports.US_GOV = US_GOV;
exports.US_GOV_DOD = US_GOV_DOD;
exports.cloudFromName = cloudFromName;
exports.withOverrides = withOverrides;
