// @microsoft/teams.apps/dist/activity-sender.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ActivitySender = void 0;
const teams_api_1 = require("@microsoft/teams.api");
const http_stream_1 = require("./http/http-stream");
/**
 * Handles sending activities to the Bot Framework
 * Separate from transport concerns (HTTP, WebSocket, etc.)
 */
class ActivitySender {
    client;
    logger;
    constructor(client, logger) {
        this.client = client;
        this.logger = logger;
    }
    async send(activity, ref) {
        // Create API client for this conversation's service URL
        const api = new teams_api_1.Client(ref.serviceUrl, this.client);
        // Merge activity with conversation reference
        activity = {
            ...activity,
            from: ref.bot,
            conversation: ref.conversation,
        };
        // Check if this is a targeted message
        const isTargeted = activity.recipient?.isTargeted === true;
        if (isTargeted && ref.conversation.conversationType === 'personal') {
            throw new Error('Targeted messages are not supported in 1:1 (personal) chats.');
        }
        // Decide create vs update, with targeted variants
        if (activity.id) {
            const res = isTargeted
                ? await api.conversations.activities(ref.conversation.id).updateTargeted(activity.id, activity)
                : await api.conversations.activities(ref.conversation.id).update(activity.id, activity);
            return { ...activity, ...res };
        }
        const res = isTargeted
            ? await api.conversations.activities(ref.conversation.id).createTargeted(activity)
            : await api.conversations.activities(ref.conversation.id).create(activity);
        return { ...activity, ...res };
    }
    createStream(ref) {
        // Create API client for this conversation's service URL
        const api = new teams_api_1.Client(ref.serviceUrl, this.client);
        return new http_stream_1.HttpStream(api, ref, this.logger);
    }
}
exports.ActivitySender = ActivitySender;
