// @microsoft/teams.apps/dist/app.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.App = void 0;
const axios_1 = require("axios");
const teams_api_1 = require("@microsoft/teams.api");
const teams_common_1 = require("@microsoft/teams.common");
const package_json_1 = __importDefault(require("../package.json"));
const activity_sender_1 = require("./activity-sender");
const api_1 = require("./api");
const app_embed_1 = require("./app.embed");
const app_events_1 = require("./app.events");
const app_oauth_1 = require("./app.oauth");
const app_plugins_1 = require("./app.plugins");
const app_process_1 = require("./app.process");
const app_routing_1 = require("./app.routing");
const container_1 = require("./container");
const http_1 = require("./http");
const http_server_1 = require("./http/http-server");
const middleware = __importStar(require("./middleware"));
const oauth_1 = require("./oauth");
const router_1 = require("./router");
const token_manager_1 = require("./token-manager");
const thread_1 = require("./utils/thread");
/**
 * The orchestrator for receiving/sending activities
 */
class App {
    options;
    api;
    cloud;
    graph;
    log;
    server;
    http;
    client;
    storage;
    entraTokenValidator;
    tokenManager;
    /**
     * Graph API base URL derived from the configured cloud's `graphScope`.
     * Undefined when the scope isn't a URL — `GraphClient` then uses its public-cloud default.
     * Shared across every `GraphClient` the app constructs (`app.graph`, `ctx.appGraph`, `ctx.userGraph`)
     * so sovereign customers get consistent routing.
     */
    graphBaseUrl;
    /**
     * the apps credentials
     */
    get credentials() {
        return this.tokenManager.credentials;
    }
    /**
     * the apps id
     */
    get id() {
        return this.credentials?.clientId;
    }
    get oauth() {
        return {
            ...oauth_1.DEFAULT_OAUTH_SETTINGS,
            ...this.options.oauth,
        };
    }
    container = new container_1.Container();
    plugins = [];
    router = new router_1.Router();
    tenantTokens = new teams_common_1.LocalStorage({}, { max: 20000 });
    events = new teams_common_1.EventEmitter();
    isInitialized = false;
    port;
    activitySender;
    _userAgent = `teams.ts[apps]/${package_json_1.default.version}`;
    constructor(options = {}) {
        this.options = options;
        this.log = this.options.logger || new teams_common_1.ConsoleLogger('@teams/app');
        this.storage = this.options.storage || new teams_common_1.LocalStorage();
        // Resolve cloud environment from options or CLOUD env var
        const cloudEnvName = typeof process !== 'undefined' ? process.env.CLOUD : undefined;
        this.cloud = this.options.cloud ?? (cloudEnvName ? (0, teams_api_1.cloudFromName)(cloudEnvName) : teams_api_1.PUBLIC);
        if (!options.client) {
            this.client = new teams_common_1.Client({
                headers: {
                    'User-Agent': this._userAgent,
                },
            });
        }
        else if (typeof options.client === 'function') {
            this.client = options.client().clone({
                headers: {
                    'User-Agent': this._userAgent,
                },
            });
        }
        else if ('request' in options.client) {
            this.client = options.client.clone({
                headers: {
                    'User-Agent': this._userAgent,
                },
            });
        }
        else {
            this.client = new teams_common_1.Client(options.client).clone({
                headers: {
                    'User-Agent': this._userAgent,
                },
            });
        }
        const serviceUrl = (this.options.serviceUrl ?? process.env.SERVICE_URL ??
            'https://smba.trafficmanager.net/teams').replace(/\/+$/, '');
        this.api = new api_1.ApiClient(serviceUrl, this.client.clone({ token: () => this.getBotToken() }), this.options.apiClientSettings, this.cloud);
        // Derive Graph API base URL from the cloud's graphScope (e.g. "https://graph.microsoft.us/.default"
        // -> "https://graph.microsoft.us"). Falls back to the public Graph endpoint inside GraphClient if
        // the scope isn't a URL (custom delegated scope, empty, etc.).
        const graphUrlMatch = /^(https?:\/\/[^/]+)/i.exec((this.cloud.graphScope ?? '').trim());
        this.graphBaseUrl = graphUrlMatch?.[1];
        if (!this.graphBaseUrl && this.cloud.graphScope) {
            this.log.warn(`graphScope "${this.cloud.graphScope}" is not a URL; Graph calls will route to the public cloud. ` +
                'Set graphScope to an "https://<host>/.default" value to route to the correct Graph endpoint.');
        }
        this.graph = new api_1.GraphClient(this.client.clone({ token: () => this.getAppGraphToken() }), { baseUrlRoot: this.graphBaseUrl });
        // initialize TokenManager with credentials
        this.tokenManager = new token_manager_1.TokenManager({
            clientId: this.options.clientId,
            clientSecret: this.options.clientSecret,
            tenantId: this.options.tenantId,
            token: this.options.token,
            managedIdentityClientId: this.options.managedIdentityClientId,
            cloud: this.cloud,
        }, this.log);
        // initialize ActivitySender for sending activities
        this.activitySender = new activity_sender_1.ActivitySender(this.client.clone({ token: () => this.getBotToken() }), this.log);
        if (this.credentials?.clientId) {
            this.entraTokenValidator = middleware.createEntraTokenValidator(this.credentials.tenantId || 'common', this.credentials.clientId, { applicationIdUri: this.options.applicationIdUri, loginEndpoint: this.cloud.loginEndpoint, logger: this.log });
        }
        // Determine HTTP server
        const plugins = this.options.plugins || [];
        const httpPlugin = plugins.find((p) => {
            const meta = (0, app_plugins_1.getMetadata)(p);
            return meta.name === 'http';
        });
        // Error if both httpServerAdapter and http plugin are provided
        if (this.options.httpServerAdapter && httpPlugin) {
            throw new Error('Cannot provide both httpServerAdapter option and HttpPlugin in plugins array. ' +
                'Use either:\n' +
                '  - new App({ httpServerAdapter: new ExpressAdapter() }) (recommended)\n' +
                '  - new App({ plugins: [new HttpPlugin()] }) (deprecated)');
        }
        let server;
        // HttpPlugin in plugins array (backwards compatibility)
        if (httpPlugin) {
            this.log.warn('[DEPRECATED] HttpPlugin in plugins array will be deprecated. Use httpServerAdapter option instead:\n' +
                '  new App({ httpServerAdapter: new ExpressAdapter() })');
            this.http = httpPlugin;
            // Extract internal server and always set this.server
            server = httpPlugin.asServer?.();
            if (!server) {
                throw new Error('HttpPlugin.asServer() returned undefined');
            }
        }
        else {
            server = new http_server_1.HttpServer(this.options.httpServerAdapter ?? new http_1.ExpressAdapter(undefined, {
                logger: this.log,
                onError: (err) => this.onError({ error: err })
            }), {
                skipAuth: this.options.skipAuth,
                logger: this.log,
                messagingEndpoint: this.options.messagingEndpoint ?? '/api/messages',
            });
        }
        // Always set this.server
        this.server = server;
        // Set callback for handling activities
        server.onRequest = (event) => this.onActivity(event);
        // add injectable items to container
        this.container.register('ILogger', { useValue: this.log });
        this.container.register('IStorage', { useValue: this.storage });
        // Register HTTP server for plugins that need HTTP capabilities
        this.container.register('IHttpServer', { useValue: server });
        // Register all plugins (including HttpPlugin if using old way)
        for (const plugin of plugins) {
            this.plugin(plugin);
        }
        if (this.options.activity?.mentions?.stripText) {
            const options = this.options.activity?.mentions?.stripText;
            this.use(middleware.stripMentionsText(typeof options === 'boolean' ? {} : options));
        }
        // default event handlers
        this.router.register({
            name: 'signin.token-exchange',
            type: 'system',
            select: activity => activity.type === 'invoke' && activity.name === 'signin/tokenExchange',
            callback: ctx => this.onTokenExchange(ctx),
        });
        this.router.register({
            name: 'signin.verify-state',
            type: 'system',
            select: activity => activity.type === 'invoke' && activity.name === 'signin/verifyState',
            callback: ctx => this.onVerifyState(ctx),
        });
        this.router.register({
            name: 'signin.failure',
            type: 'system',
            select: activity => activity.type === 'invoke' && activity.name === 'signin/failure',
            callback: ctx => this.onSignInFailure(ctx),
        });
        this.event('error', ({ error }) => {
            this.log.error(error.message);
            if (error instanceof axios_1.AxiosError) {
                this.log.error(error.request.path);
                this.log.error(error.response?.data);
            }
        });
    }
    /**
     * initialize the app.
     */
    async initialize() {
        if (this.isInitialized) {
            return;
        }
        // initialize plugins
        for (const plugin of this.plugins) {
            this.inject(plugin);
            if (plugin.onInit) {
                await plugin.onInit();
            }
        }
        // initialize server
        await this.server.initialize({
            credentials: this.credentials,
            cloud: this.cloud,
        });
        this.isInitialized = true;
    }
    /**
     * start the server after initialization
     * @param port port to listen on
     */
    async start(port) {
        this.port = port || process.env.PORT || 3978;
        try {
            await this.initialize();
            // Start plugins
            for (const plugin of this.plugins) {
                if (plugin.onStart) {
                    await plugin.onStart({ port: this.port });
                }
            }
            this.events.emit('start', this.log);
            // Start HTTP server
            await this.server.start(this.port);
        }
        catch (error) {
            await this.stop();
            this.onError({ error });
        }
    }
    /**
     * stop the app
     */
    async stop() {
        try {
            // Stop plugins
            for (const plugin of this.plugins) {
                if (plugin.onStop) {
                    await plugin.onStop();
                }
            }
            // Stop HTTP server
            await this.server.stop();
        }
        catch (error) {
            this.onError({ error });
        }
    }
    /**
     * send an activity proactively to a conversation.
     *
     * Sends to the exact conversation ID provided. For channel threads,
     * the conversation ID must include `;messageid=` - use {@link toThreadedConversationId}
     * to construct it, or use {@link reply} which handles this automatically.
     *
     * @param conversationId the conversation to send to
     * @param activity the activity to send
     */
    async send(conversationId, activity) {
        if (!this.id) {
            throw new Error('App has no credentials set up');
        }
        const params = (0, teams_api_1.toActivityParams)(activity);
        const ref = {
            channelId: 'msteams',
            serviceUrl: this.api.serviceUrl,
            bot: {
                id: this.id,
                role: 'bot',
            },
            conversation: {
                id: conversationId,
            },
        };
        const res = await this.activitySender.send(params, ref);
        return res;
    }
    async reply(conversationId, messageId, activity) {
        if (typeof messageId === 'string' && activity !== undefined) {
            return this.send((0, thread_1.toThreadedConversationId)(conversationId, messageId), activity);
        }
        return this.send(conversationId, messageId);
    }
    /**
     * subscribe to an event
     * @param name event to subscribe to
     * @param cb callback to invoke
     */
    on = app_routing_1.on; // eslint-disable-line @typescript-eslint/member-ordering
    /**
     * subscribe to a message event for a specific pattern
     * @param pattern pattern to match against message text
     * @param cb callback to invoke
     */
    message = app_routing_1.message; // eslint-disable-line @typescript-eslint/member-ordering
    /**
     * register a middleware
     * @param cb callback to invoke
     */
    use = app_routing_1.use; // eslint-disable-line @typescript-eslint/member-ordering
    /**
     * subscribe to an event
     * @param name the event to subscribe to
     * @param cb the callback to invoke
     */
    event = app_events_1.event; // eslint-disable-line @typescript-eslint/member-ordering
    /**
     * add a plugin
     * @param plugin plugin to add
     */
    plugin = app_plugins_1.plugin; // eslint-disable-line @typescript-eslint/member-ordering
    /**
     * get a plugin
     */
    getPlugin = app_plugins_1.getPlugin; // eslint-disable-line @typescript-eslint/member-ordering
    /**
     * add/update a function that can be called remotely
     * @param name The unique function name
     * @param cb The callback to handle the function
     */
    function = app_embed_1.func; // eslint-disable-line @typescript-eslint/member-ordering
    /**
     * add/update a static tab.
     * the tab will be hosted at
     * `http://localhost:{{PORT}}/tabs/{{name}}` or `https://{{BOT_DOMAIN}}/tabs/{{name}}`
     * @remark scopes default to `personal`
     * @param name A unique identifier for the entity which the tab displays.
     * @param path The path to the web `dist` folder.
     */
    tab = app_embed_1.tab; // eslint-disable-line @typescript-eslint/member-ordering
    /**
     * activity handler called when an inbound activity is received
     * @param sender the plugin to use for sending activities
     * @param event the received activity event
     */
    process = app_process_1.$process; // eslint-disable-line @typescript-eslint/member-ordering
    ///
    /// OAuth
    ///
    onTokenExchange = app_oauth_1.onTokenExchange; // eslint-disable-line @typescript-eslint/member-ordering
    onVerifyState = app_oauth_1.onVerifyState; // eslint-disable-line @typescript-eslint/member-ordering
    onSignInFailure = app_oauth_1.onSignInFailure; // eslint-disable-line @typescript-eslint/member-ordering
    ///
    /// Events
    ///
    inject = app_plugins_1.inject; // eslint-disable-line @typescript-eslint/member-ordering
    onError = app_events_1.onError; // eslint-disable-line @typescript-eslint/member-ordering
    onActivitySent = app_events_1.onActivitySent; // eslint-disable-line @typescript-eslint/member-ordering
    onActivityResponse = app_events_1.onActivityResponse; // eslint-disable-line @typescript-eslint/member-ordering
    async onActivity(event) {
        this.events.emit('activity', event);
        return await this.process(event);
    }
    ///
    /// Token
    ///
    async getBotToken() {
        if (!this.tokenManager)
            return;
        return await this.tokenManager.getBotToken();
    }
    async getUserToken(channelId, userId) {
        const res = await this.api.users.token.get({
            channelId,
            userId,
            connectionName: this.oauth.defaultConnectionName,
        });
        return res.token;
    }
    async getAppGraphToken(tenantId) {
        if (!this.tokenManager)
            return;
        return await this.tokenManager.getGraphToken(tenantId);
    }
}
exports.App = App;
