// @microsoft/teams.client/dist/app.mjs
import { createNestablePublicClientApplication } from '@azure/msal-browser';
import { app } from '@microsoft/teams-js';
import { ConsoleLogger, Client } from '@microsoft/teams.common';
import { buildGraphClient } from './graph-utils.mjs';
import { buildMsalConfig, acquireMsalAccessToken, getStandardExecSilentRequest, hasConsentForScopes } from './msal-utils.mjs';

class App {
  options;
  http;
  graph;
  clientId;
  _state = { phase: "stopped" };
  /**
   * The apps logger
   */
  get log() {
    return this._log;
  }
  _log;
  /**
   * The date/time when the app was successfully started.
   */
  get startedAt() {
    return this._state?.startedAt;
  }
  /** The msal instance used in this app. undefined until the app is started. */
  get msalInstance() {
    return this._state.msalInstance;
  }
  constructor(clientId, options = {}) {
    if (!clientId) {
      throw new Error("Invalid client ID.");
    }
    this.clientId = clientId;
    this.options = options;
    this._log = options?.logger || new ConsoleLogger("@teams/client");
    this.http = new Client({
      baseUrl: options?.remoteApiOptions?.baseUrl
    });
    this.graph = buildGraphClient(
      () => this.appStateGuard(),
      this._log,
      this.options.msalOptions?.prewarmScopes ? () => this.options.msalOptions.prewarmScopes : void 0
    );
  }
  /**
   * Starts the library and initializes the dependent teams-js and MSAL libraries.
   * @returns A promise that will be fulfilled when the app has started, or
   *          rejected if the initialization fails or times out.
   */
  async start() {
    if (this._state.phase !== "stopped") {
      this._log.debug(`app already ${this._state.phase}`);
      return;
    }
    this._log.debug("app starting");
    this._state = { phase: "starting" };
    await app.initialize();
    let msalInstance = this.options.msalOptions?.msalInstance;
    if (!msalInstance) {
      const msalConfig = this.options.msalOptions?.configuration ?? buildMsalConfig(this.clientId, this._log);
      msalInstance = await createNestablePublicClientApplication(
        msalConfig
      );
    }
    this._state = { phase: "started", msalInstance, startedAt: /* @__PURE__ */ new Date() };
    if (this.options.msalOptions?.prewarmScopes !== false) {
      const scopes = this.options.msalOptions?.prewarmScopes ?? [".default"];
      this._log.debug(`prewarming consent for scopes: ${scopes.join(", ")}`);
      await this.ensureConsentForScopes(scopes);
    }
    this._log.debug("app started");
  }
  /**
   * Execute a server-side function
   * @param name The unique function name
   * @param data The data to send
   * @param options Options
   * @param options.msalTokenRequest Optional MSAL token request.
   * If omitted, a default token request is used.
   * @param options.requestHeaders Optional additional request headers.
   * @returns The function response
   */
  async exec(name, data, options) {
    const { msalInstance } = this.appStateGuard();
    const context = await app.getContext();
    const remoteAppResource = this.options.remoteApiOptions?.remoteAppResource ?? `api://${this.clientId}`;
    const accessToken = await acquireMsalAccessToken(
      msalInstance,
      options?.msalTokenRequest ?? getStandardExecSilentRequest(remoteAppResource, options?.permission),
      this._log
    );
    const res = await this.http.post(`/api/functions/${name}`, data, {
      headers: {
        authorization: `Bearer ${accessToken}`,
        "x-teams-app-session-id": context.app.sessionId,
        "x-teams-channel-id": context.channel?.id,
        "x-teams-chat-id": context.chat?.id,
        "x-teams-meeting-id": context.meeting?.id,
        "x-teams-message-id": context.app.parentMessageId || void 0,
        "x-teams-page-id": context.page.id || void 0,
        "x-teams-sub-page-id": context.page.subPageId || void 0,
        "x-teams-team-id": context.team?.internalId,
        ...options?.requestHeaders ?? {}
      }
    });
    return res.data;
  }
  /**
   * Tests whether the user has consented to the specified scopes without prompting the user for consent.
   * @param scopes The scopes to check consent for.The scopes should be for a single resource, and they
   * should not mix the .default scope with named scopes.
   * @returns A promise that resolves to a boolean indicating whether the user has consented to the scopes.
   */
  async hasConsentForScopes(scopes) {
    const { msalInstance } = this.appStateGuard();
    return await hasConsentForScopes(msalInstance, scopes, this.log);
  }
  /**
   * Tests whether the user has consented to the specified scopes, and prompts them if not. This is useful for ensuring
   * that the user has consented to the required scopes before calling a graph API or other resource.
   * @param scopes - The scopes to prewarm consent for. The scopes should be for a single resource, and they
   * should not mix the .default scope with named scopes.
   * @returns A value indicating whether consent has been acquired for the specified scopes.
   */
  async ensureConsentForScopes(scopes) {
    const { msalInstance } = this.appStateGuard();
    try {
      const token = await acquireMsalAccessToken(
        msalInstance,
        { scopes },
        this.log
      );
      return !!token;
    } catch (ex) {
      return false;
    }
  }
  appStateGuard() {
    if (this._state.phase !== "started") {
      throw new Error("App not started");
    }
    return this._state;
  }
}

export { App };
