// @microsoft/teams.client/dist/msal-utils.mjs
import { LogLevel, InteractionRequiredAuthError } from '@azure/msal-browser';

const shouldTryPopup = (ex) => {
  if (ex instanceof InteractionRequiredAuthError) {
    return true;
  }
  const errorCode = ex?.errorCode ?? "";
  if (errorCode === "ApiContractViolation") {
    return true;
  }
  return false;
};
const getStandardExecSilentRequest = (resource, permission = "access_as_user") => ({
  scopes: [`${resource}/${permission}`]
});
const buildMsalConfig = (clientId, logger) => {
  return {
    auth: {
      clientId,
      supportsNestedAppAuth: true,
      redirectUri: "/",
      postLogoutRedirectUri: "/"
    },
    system: {
      loggerOptions: {
        piiLoggingEnabled: false,
        loggerCallback: (level, message) => {
          switch (level) {
            case LogLevel.Error:
              logger.error(message);
              return;
            case LogLevel.Info:
              logger.info(message);
              return;
            case LogLevel.Verbose:
              logger.debug(message);
              return;
            case LogLevel.Warning:
              logger.warn(message);
              return;
            default:
              return;
          }
        }
      }
    }
  };
};
const acquireMsalAccessToken = async (msalInstance, request, logger) => {
  try {
    const response = await msalInstance.acquireTokenSilent(request);
    return response.accessToken;
  } catch (ex) {
    if (!shouldTryPopup(ex)) {
      logger.error("acquireTokenSilent failed", ex);
      throw ex;
    }
  }
  try {
    logger.debug("acquireTokenSilent failed; trying acquireTokenPopup");
    const response = await msalInstance.acquireTokenPopup(request);
    return response.accessToken;
  } catch (ex) {
    const errorCode = ex?.errorCode ?? "";
    if (errorCode === "ApiContractViolation") {
      logger.error(
        "acquireTokenPopup failed with ApiContractViolation. On Teams Desktop, the OneAuth broker cannot resolve the '.default' scope. Set explicit scopes in msalOptions.prewarmScopes (e.g., ['User.Read']) for Desktop compatibility.",
        ex
      );
    } else {
      logger.error("acquireTokenPopup failed", ex);
    }
    throw ex;
  }
};
const hasConsentForScopes = async (msalInstance, scopes, logger) => {
  try {
    await msalInstance.acquireTokenSilent({
      scopes
    });
    return true;
  } catch (ex) {
    const acquireTokenPopupNeeded = ex instanceof InteractionRequiredAuthError;
    const logLevel = acquireTokenPopupNeeded ? "debug" : "error";
    logger.log(logLevel, "hasConsentForScopes failed", ex);
    return false;
  }
};

export { acquireMsalAccessToken, buildMsalConfig, getStandardExecSilentRequest, hasConsentForScopes };
