// @microsoft/teams.apps/dist/contexts/activity.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ActivityContext = void 0;
const teams_api_1 = require("@microsoft/teams.api");
class ActivityContext {
    appId;
    activity;
    ref;
    log;
    api;
    appGraph;
    userGraph;
    storage;
    stream;
    isSignedIn;
    connectionName;
    next;
    activitySender;
    constructor(value) {
        // Extract activitySender and next before Object.assign to avoid overwriting methods
        const { activitySender, next, ...rest } = value;
        if (rest.activity.type === 'message') {
            rest.activity = teams_api_1.MessageActivity.from(rest.activity).toInterface();
        }
        if (rest.activity.type === 'messageUpdate') {
            rest.activity = teams_api_1.MessageUpdateActivity.from(rest.activity).toInterface();
        }
        if (rest.activity.type === 'messageDelete') {
            rest.activity = teams_api_1.MessageDeleteActivity.from(rest.activity).toInterface();
        }
        if (rest.activity.type === 'typing') {
            rest.activity = teams_api_1.TypingActivity.from(rest.activity).toInterface();
        }
        // SECURITY: drop any keys in `rest` that would shadow prototype methods.
        // Plugin-supplied context can add new properties via the [key: string]: any
        // index signature, but must not overwrite methods that callers rely on for
        // trust (send, reply, quote, signin, signout).
        for (const key of PROTECTED_METHOD_NAMES) {
            if (key in rest) {
                delete rest[key];
            }
        }
        Object.assign(this, rest);
        this.activitySender = activitySender;
        this.next = next;
        this.stream = activitySender.createStream(value.ref);
        this.connectionName = value.connectionName;
    }
    /**
     * send an activity in the current conversation without quoting.
     *
     * In channels, sends to the current thread. In scopes that do not
     * support threading (group chat, meetings), sends as a normal message.
     * To send with a visual quote of the inbound message, use {@link reply}.
     *
     * @param activity the activity to send
     * @param conversationRef optional conversation reference to send to a different conversation or thread
     */
    async send(activity, conversationRef) {
        const params = (0, teams_api_1.toActivityParams)(activity);
        if (this.shouldOutboundBeAutoTargeted(params, conversationRef)) {
            this.applyTargetedRecipient(params);
        }
        if (this.isTargetedOutbound(params)) {
            this.stripQuotedReplyMetadata(params);
            // `targetedMessageInfo` points at the original targeted inbound message for prompt preview.
            // Do not add it for generic targeted sends; Teams can reject it if the referenced activity
            // was not itself delivered as a targeted message.
            if (this.isIncomingTargeted()) {
                this.addTargetedMessageInfo(params);
            }
        }
        return await this.activitySender.send(params, conversationRef ?? this.ref);
    }
    /**
     * send an activity in the current conversation with a visual quote
     * of the inbound message.
     *
     * In channels, sends to the current thread with a quoted reply.
     * In other scopes, sends with a quoted reply.
     * To send without quoting, use {@link send}.
     *
     * @param activity the activity to send
     */
    async reply(activity) {
        if (this.activity.id) {
            return this.quote(this.activity.id, activity);
        }
        return this.send(activity);
    }
    /**
     * Send a message to the conversation with a quoted message reference prepended to the text.
     * Teams renders the quoted message as a preview bubble above the response text.
     * @param messageId - The ID of the message to quote
     * @param activity - The activity to send — a quote placeholder for messageId will be prepended to its text
     *
     * @experimental This API is coming soon and may change in the future.
     * Diagnostic: ExperimentalTeamsQuotedReplies
     */
    async quote(messageId, activity) {
        activity = (0, teams_api_1.toActivityParams)(activity);
        if (activity.type === 'message') {
            const message = teams_api_1.MessageActivity.from(activity);
            message.prependQuote(messageId);
            return this.send(message);
        }
        return this.send(activity);
    }
    async signin(options) {
        const { oauthCardText, signInButtonText, connectionName, signInLink, overrideSignInActivity } = {
            oauthCardText: 'Please Sign In...',
            signInButtonText: 'Sign In',
            ...options,
        };
        const convo = { ...this.ref };
        try {
            const res = await this.api.users.token.get({
                channelId: this.activity.channelId,
                userId: this.activity.from.id,
                connectionName: connectionName || this.connectionName,
            });
            return res.token;
        }
        catch (err) {
            // noop
        }
        const tokenExchangeState = {
            connectionName: connectionName || this.connectionName,
            conversation: convo,
            relatesTo: this.activity.relatesTo,
            msAppId: this.appId,
        };
        if (this.activity.conversation.isGroup) {
            // create new 1:1 conversation with user to do SSO
            // because groupchats don't support it.
            const res = await this.api.conversations.create({
                tenantId: this.activity.conversation.tenantId,
                members: [this.activity.from],
            });
            await this.send({ type: 'message', text: oauthCardText });
            convo.conversation = { id: res.id };
        }
        const state = Buffer.from(JSON.stringify(tokenExchangeState)).toString('base64');
        const resource = await this.api.bots.signIn.getResource({ state });
        await this.send(overrideSignInActivity?.(resource.tokenExchangeResource, resource.tokenPostResource, resource.signInLink) ?? {
            type: 'message',
            inputHint: 'acceptingInput',
            recipient: this.activity.from,
            conversation: convo.conversation,
            attachments: [
                (0, teams_api_1.cardAttachment)('oauth', {
                    text: oauthCardText,
                    connectionName: connectionName || this.connectionName,
                    tokenExchangeResource: resource.tokenExchangeResource,
                    tokenPostResource: resource.tokenPostResource,
                    buttons: [
                        {
                            type: 'signin',
                            title: signInButtonText,
                            value: signInLink || resource.signInLink,
                        },
                    ],
                }),
            ],
        }, convo);
    }
    async signout(connectionName) {
        await this.api.users.token.signOut({
            channelId: this.activity.channelId,
            userId: this.activity.from.id,
            connectionName: connectionName || this.connectionName,
        });
    }
    toInterface() {
        return {
            activity: this.activity,
            api: this.api,
            appGraph: this.appGraph,
            userGraph: this.userGraph,
            appId: this.appId,
            log: this.log,
            ref: this.ref,
            storage: this.storage,
            stream: this.stream,
            isSignedIn: this.isSignedIn,
            connectionName: this.connectionName,
            userToken: this.userToken,
            next: this.next.bind(this),
            reply: this.reply.bind(this),
            quote: this.quote.bind(this),
            send: this.send.bind(this),
            signin: this.signin.bind(this),
            signout: this.signout.bind(this),
        };
    }
    isIncomingTargeted() {
        return this.activity.recipient?.isTargeted === true;
    }
    shouldOutboundBeAutoTargeted(params, conversationRef) {
        if (params.type !== 'message') {
            return false;
        }
        if (!this.isIncomingTargeted()) {
            return false;
        }
        if (!this.isSameConversation(conversationRef)) {
            return false;
        }
        return !params.id && !params.recipient;
    }
    isSameConversation(conversationRef) {
        return !conversationRef || conversationRef.conversation?.id === this.ref.conversation?.id;
    }
    applyTargetedRecipient(params) {
        params.recipient = {
            ...this.activity.from,
            isTargeted: true,
        };
    }
    isTargetedOutbound(params) {
        return params.type === 'message' && params.recipient?.isTargeted === true;
    }
    stripQuotedReplyMetadata(params) {
        if (params.entities) {
            params.entities = params.entities.filter((e) => e.type !== 'quotedReply');
        }
        if (params.text) {
            params.text = params.text.replace(`<quoted messageId="${this.activity.id}"/>`, '').trim();
        }
    }
    addTargetedMessageInfo(params) {
        if (params.entities?.some((e) => e.type === 'targetedMessageInfo')) {
            return;
        }
        if (!params.entities) {
            params.entities = [];
        }
        params.entities.push({
            type: 'targetedMessageInfo',
            messageId: this.activity.id,
        });
    }
}
exports.ActivityContext = ActivityContext;
// Names of prototype methods (and getters) on ActivityContext that must not be
// shadowed by instance properties from external context. Computed once from the
// prototype chain at module load, so any method or accessor added to the class
// in the future is protected automatically with no maintenance.
const PROTECTED_METHOD_NAMES = (() => {
    const names = new Set();
    let proto = ActivityContext.prototype;
    while (proto && proto !== Object.prototype) {
        for (const name of Object.getOwnPropertyNames(proto)) {
            if (name === 'constructor')
                continue;
            const descriptor = Object.getOwnPropertyDescriptor(proto, name);
            if (!descriptor)
                continue;
            if (typeof descriptor.value === 'function' || typeof descriptor.get === 'function') {
                names.add(name);
            }
        }
        proto = Object.getPrototypeOf(proto);
    }
    return names;
})();
