// @microsoft/teams.apps/dist/contexts/activity.test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const teams_api_1 = require("@microsoft/teams.api");
const activity_1 = require("./activity");
describe('ActivityContext', () => {
    let mockSender;
    let mockApiClient;
    let mockLogger;
    let mockStorage;
    let mockRef;
    let context;
    beforeEach(() => {
        mockSender = {
            createStream: jest.fn().mockReturnValue({ write: jest.fn() }),
            send: jest.fn(),
        };
        mockApiClient = {
            users: {
                token: {
                    get: jest.fn(),
                    signOut: jest.fn(),
                },
            },
            conversations: {
                create: jest.fn(),
            },
            bots: {
                signIn: {
                    getResource: jest.fn(),
                },
            },
        };
        mockStorage = {};
        const partialLogger = {
            debug: jest.fn(),
            info: jest.fn(),
            warn: jest.fn(),
            error: jest.fn(),
            trace: jest.fn(),
            log: jest.fn(),
        };
        mockLogger = {
            ...partialLogger,
            child: jest.fn().mockReturnValue(partialLogger),
        };
        mockRef = {
            activityId: 'test-activity-id',
            channelId: 'test-channel',
            conversation: {
                id: 'test-conversation',
                conversationType: 'channel',
                isGroup: false,
            },
            serviceUrl: 'https://test.com',
            user: { id: 'test-user', name: 'Test User', role: 'user' },
            bot: { id: 'bot-id', name: 'Bot', role: 'bot' },
        };
    });
    const buildIncomingMessageActivity = (text, id) => {
        return new teams_api_1.MessageActivity(text)
            .withFrom({
            id: 'test-user',
            name: 'Test User',
            role: 'user',
        })
            .withRecipient({
            id: 'bot-id',
            name: 'Bot',
            role: 'bot',
        })
            .withChannelId('test-channel')
            .withConversation({
            id: 'test-conversation',
            conversationType: 'channel',
            isGroup: false,
        })
            .withId(id ?? 'test-activity-id');
    };
    const buildActivityContext = (activity) => {
        return new activity_1.ActivityContext({
            appId: 'test-app',
            activity,
            ref: mockRef,
            log: mockLogger,
            api: mockApiClient,
            appGraph: {},
            userGraph: {},
            storage: mockStorage,
            connectionName: 'test-connection',
            next: jest.fn(),
            activitySender: mockSender,
        });
    };
    describe('reply', () => {
        it('stamps quotedReply entity with activity id', async () => {
            const activity = buildIncomingMessageActivity('Hello world');
            context = buildActivityContext(activity);
            await context.reply('What is up?');
            expect(mockSender.send).toHaveBeenCalledTimes(1);
            const sentActivity = mockSender.send.mock.calls[0][0];
            expect(sentActivity.entities).toEqual(expect.arrayContaining([
                expect.objectContaining({
                    type: 'quotedReply',
                    quotedReply: { messageId: 'test-activity-id' },
                }),
            ]));
        });
        it('prepends placeholder to text', async () => {
            const activity = buildIncomingMessageActivity('Hello world');
            context = buildActivityContext(activity);
            await context.reply('What is up?');
            const sentActivity = mockSender.send.mock.calls[0][0];
            expect(sentActivity.text).toEqual('<quoted messageId="test-activity-id"/> What is up?');
        });
        it('sets placeholder as text when reply text is empty', async () => {
            const activity = buildIncomingMessageActivity('Hello world');
            context = buildActivityContext(activity);
            await context.reply('');
            const sentActivity = mockSender.send.mock.calls[0][0];
            expect(sentActivity.text).toEqual('<quoted messageId="test-activity-id"/>');
        });
        it('sets placeholder as text when reply has no text', async () => {
            const activity = buildIncomingMessageActivity('Hello world');
            context = buildActivityContext(activity);
            await context.reply({ type: 'message' });
            const sentActivity = mockSender.send.mock.calls[0][0];
            expect(sentActivity.text).toEqual('<quoted messageId="test-activity-id"/>');
        });
        it('does not stamp entity when activity has no id', async () => {
            const activity = buildIncomingMessageActivity('Hello world', '');
            context = buildActivityContext(activity);
            await context.reply('What is up?');
            const sentActivity = mockSender.send.mock.calls[0][0];
            expect(sentActivity.entities).toBeUndefined();
        });
    });
    describe('quote', () => {
        it('stamps quotedReply entity with given messageId', async () => {
            const activity = buildIncomingMessageActivity('Hello world');
            context = buildActivityContext(activity);
            await context.quote('arbitrary-msg-id', 'some text');
            expect(mockSender.send).toHaveBeenCalledTimes(1);
            const sentActivity = mockSender.send.mock.calls[0][0];
            expect(sentActivity.entities).toEqual(expect.arrayContaining([
                expect.objectContaining({
                    type: 'quotedReply',
                    quotedReply: { messageId: 'arbitrary-msg-id' },
                }),
            ]));
        });
        it('prepends placeholder to text', async () => {
            const activity = buildIncomingMessageActivity('Hello world');
            context = buildActivityContext(activity);
            await context.quote('msg-42', 'reply text');
            const sentActivity = mockSender.send.mock.calls[0][0];
            expect(sentActivity.text).toEqual('<quoted messageId="msg-42"/> reply text');
        });
        it('sets placeholder as text when no text provided', async () => {
            const activity = buildIncomingMessageActivity('Hello world');
            context = buildActivityContext(activity);
            await context.quote('msg-42', { type: 'message' });
            const sentActivity = mockSender.send.mock.calls[0][0];
            expect(sentActivity.text).toEqual('<quoted messageId="msg-42"/>');
        });
        it('reply to targeted message strips blockquote via addTargetedMessageInfo', async () => {
            const activity = new teams_api_1.MessageActivity('Hello world')
                .withFrom({ id: 'test-user', name: 'Test User', role: 'user' })
                .withRecipient({ id: 'bot-id', name: 'Bot', role: 'bot' }, true)
                .withChannelId('test-channel')
                .withConversation({ id: 'test-conversation', conversationType: 'channel', isGroup: false })
                .withId('test-activity-id');
            context = buildActivityContext(activity);
            await context.reply('Here is your agenda');
            expect(mockSender.send).toHaveBeenCalledTimes(1);
            const sentActivity = mockSender.send.mock.calls[0][0];
            // Reply prepends blockquote, but send() auto-populates addTargetedMessageInfo
            // which strips quotedReply entities — the blockquote text remains since it's
            // the legacy format, not the <quoted .../> placeholder.
            expect(sentActivity.recipient).toEqual(expect.objectContaining({
                id: 'test-user',
                name: 'Test User',
                role: 'user',
                isTargeted: true,
            }));
            expect(sentActivity.entities).toEqual(expect.arrayContaining([
                expect.objectContaining({
                    type: 'targetedMessageInfo',
                    messageId: 'test-activity-id',
                }),
            ]));
        });
    });
    describe('send', () => {
        it('sends the activity to the sender plugin', async () => {
            const activity = buildIncomingMessageActivity('Hello world');
            context = buildActivityContext(activity);
            await context.send('What is up?');
            expect(mockSender.send).toHaveBeenCalledTimes(1);
            expect(mockSender.send).toHaveBeenCalledWith(expect.objectContaining({
                text: 'What is up?',
                type: 'message',
            }), mockRef);
        });
        describe('targeted messages', () => {
            it('defaults send to targeted when inbound message is targeted', async () => {
                const activity = new teams_api_1.MessageActivity('Hello world')
                    .withFrom({ id: 'test-user', name: 'Test User', role: 'user' })
                    .withRecipient({ id: 'bot-id', name: 'Bot', role: 'bot' }, true)
                    .withChannelId('test-channel')
                    .withConversation({ id: 'test-conversation', conversationType: 'channel', isGroup: false })
                    .withId('test-activity-id');
                context = buildActivityContext(activity);
                await context.send('Secret message');
                expect(mockSender.send).toHaveBeenCalledTimes(1);
                expect(mockSender.send).toHaveBeenCalledWith(expect.objectContaining({
                    text: 'Secret message',
                    type: 'message',
                    recipient: expect.objectContaining({ id: 'test-user', name: 'Test User', role: 'user', isTargeted: true }),
                }), mockRef);
            });
            it('does not default send to targeted for a different conversation', async () => {
                const activity = new teams_api_1.MessageActivity('Hello world')
                    .withFrom({ id: 'test-user', name: 'Test User', role: 'user' })
                    .withRecipient({ id: 'bot-id', name: 'Bot', role: 'bot' }, true)
                    .withChannelId('test-channel')
                    .withConversation({ id: 'test-conversation', conversationType: 'channel', isGroup: false })
                    .withId('test-activity-id');
                context = buildActivityContext(activity);
                const otherRef = {
                    ...mockRef,
                    conversation: {
                        ...mockRef.conversation,
                        id: 'other-conversation',
                    },
                };
                await context.send('Secret message', otherRef);
                expect(mockSender.send).toHaveBeenCalledTimes(1);
                expect(mockSender.send).toHaveBeenCalledWith(expect.objectContaining({
                    text: 'Secret message',
                    type: 'message',
                }), otherRef);
                const sentActivity = mockSender.send.mock.calls[0][0];
                expect(sentActivity.recipient).toBeUndefined();
                expect(sentActivity.entities).toBeUndefined();
            });
            it('does not default send to targeted when an explicit different recipient is supplied', async () => {
                const activity = new teams_api_1.MessageActivity('Hello world')
                    .withFrom({ id: 'test-user', name: 'Test User', role: 'user' })
                    .withRecipient({ id: 'bot-id', name: 'Bot', role: 'bot' }, true)
                    .withChannelId('test-channel')
                    .withConversation({ id: 'test-conversation', conversationType: 'channel', isGroup: false })
                    .withId('test-activity-id');
                context = buildActivityContext(activity);
                const otherRecipient = { id: 'other-user', name: 'Other User', role: 'user' };
                await context.send(new teams_api_1.MessageActivity('Public message').withRecipient(otherRecipient));
                expect(mockSender.send).toHaveBeenCalledTimes(1);
                const sentActivity = mockSender.send.mock.calls[0][0];
                expect(sentActivity).toEqual(expect.objectContaining({
                    text: 'Public message',
                    type: 'message',
                    recipient: expect.objectContaining(otherRecipient),
                }));
                expect(sentActivity.recipient.isTargeted).toBeUndefined();
                expect(sentActivity.entities).toBeUndefined();
            });
            it('sends targeted message with recipient from incoming activity', async () => {
                const activity = buildIncomingMessageActivity('Hello world');
                context = buildActivityContext(activity);
                const targetedActivity = new teams_api_1.MessageActivity('Secret message')
                    .withRecipient({ id: 'test-user', name: 'Test User', role: 'user' }, true);
                await context.send(targetedActivity);
                expect(mockSender.send).toHaveBeenCalledTimes(1);
                expect(mockSender.send).toHaveBeenCalledWith(expect.objectContaining({
                    text: 'Secret message',
                    type: 'message',
                    recipient: expect.objectContaining({ id: 'test-user', name: 'Test User', role: 'user', isTargeted: true }),
                }), mockRef);
                const sentActivity = mockSender.send.mock.calls[0][0];
                expect(sentActivity.entities).toBeUndefined();
            });
            it('allows explicitly public send from a targeted inbound message', async () => {
                const activity = new teams_api_1.MessageActivity('Hello world')
                    .withFrom({ id: 'test-user', name: 'Test User', role: 'user' })
                    .withRecipient({ id: 'bot-id', name: 'Bot', role: 'bot' }, true)
                    .withChannelId('test-channel')
                    .withConversation({ id: 'test-conversation', conversationType: 'channel', isGroup: false })
                    .withId('test-activity-id');
                context = buildActivityContext(activity);
                await context.send(new teams_api_1.MessageActivity('Public message').withRecipient(activity.from));
                expect(mockSender.send).toHaveBeenCalledTimes(1);
                const sentActivity = mockSender.send.mock.calls[0][0];
                expect(sentActivity).toEqual(expect.objectContaining({
                    text: 'Public message',
                    type: 'message',
                    recipient: expect.objectContaining({ id: 'test-user', name: 'Test User', role: 'user' }),
                }));
                expect(sentActivity.recipient.isTargeted).toBeUndefined();
                expect(sentActivity.entities).toBeUndefined();
            });
            it('sends targeted message with explicit recipient id', async () => {
                const activity = buildIncomingMessageActivity('Hello world');
                context = buildActivityContext(activity);
                const targetedActivity = new teams_api_1.MessageActivity('Secret message')
                    .withRecipient({ id: 'explicit-user-id', name: '', role: 'user' }, true);
                await context.send(targetedActivity);
                expect(mockSender.send).toHaveBeenCalledTimes(1);
                expect(mockSender.send).toHaveBeenCalledWith(expect.objectContaining({
                    text: 'Secret message',
                    type: 'message',
                    recipient: expect.objectContaining({ id: 'explicit-user-id', name: '', role: 'user', isTargeted: true }),
                }), mockRef);
            });
            it('does not set recipient for targeted message updates', async () => {
                const activity = buildIncomingMessageActivity('Hello world');
                context = buildActivityContext(activity);
                const updateActivity = new teams_api_1.MessageActivity('Updated message')
                    .withId('existing-activity-id')
                    .withRecipient({ id: 'user-1', name: '', role: 'user' }, true);
                await context.send(updateActivity);
                expect(mockSender.send).toHaveBeenCalledTimes(1);
                const sentActivity = mockSender.send.mock.calls[0][0];
                expect(sentActivity.id).toBe('existing-activity-id');
                expect(sentActivity.recipient.isTargeted).toBe(true);
            });
        });
        describe('prompt preview', () => {
            it('auto-populates targetedMessageInfo entity when incoming activity is targeted', async () => {
                const activity = new teams_api_1.MessageActivity('Hello world')
                    .withFrom({ id: 'test-user', name: 'Test User', role: 'user' })
                    .withRecipient({ id: 'bot-id', name: 'Bot', role: 'bot' }, true)
                    .withChannelId('test-channel')
                    .withConversation({ id: 'test-conversation', conversationType: 'channel', isGroup: false })
                    .withId('1772129782775');
                context = buildActivityContext(activity);
                await context.send('Here is your agenda');
                expect(mockSender.send).toHaveBeenCalledTimes(1);
                expect(mockSender.send).toHaveBeenCalledWith(expect.objectContaining({
                    text: 'Here is your agenda',
                    type: 'message',
                    entities: expect.arrayContaining([
                        expect.objectContaining({
                            type: 'targetedMessageInfo',
                            messageId: '1772129782775',
                        }),
                    ]),
                }), mockRef);
            });
            it('does not auto-populate targetedMessageInfo when incoming activity is not targeted', async () => {
                const activity = buildIncomingMessageActivity('Hello world');
                context = buildActivityContext(activity);
                await context.send('Response');
                expect(mockSender.send).toHaveBeenCalledTimes(1);
                const sentActivity = mockSender.send.mock.calls[0][0];
                expect(sentActivity.entities).toBeUndefined();
            });
            it('does not overwrite existing targetedMessageInfo entity', async () => {
                const activity = new teams_api_1.MessageActivity('Hello world')
                    .withFrom({ id: 'test-user', name: 'Test User', role: 'user' })
                    .withRecipient({ id: 'bot-id', name: 'Bot', role: 'bot' }, true)
                    .withChannelId('test-channel')
                    .withConversation({ id: 'test-conversation', conversationType: 'channel', isGroup: false })
                    .withId('1772129782775');
                context = buildActivityContext(activity);
                const outgoing = new teams_api_1.MessageActivity('Response')
                    .addTargetedMessageInfo('custom-message-id');
                await context.send(outgoing);
                expect(mockSender.send).toHaveBeenCalledTimes(1);
                const sentActivity = mockSender.send.mock.calls[0][0];
                const targetedEntities = sentActivity.entities.filter((e) => e.type === 'targetedMessageInfo');
                expect(targetedEntities).toHaveLength(1);
                expect(targetedEntities[0].messageId).toBe('custom-message-id');
            });
            it('auto-populates targetedMessageInfo on reply to targeted message', async () => {
                const activity = new teams_api_1.MessageActivity('Hello world')
                    .withFrom({ id: 'test-user', name: 'Test User', role: 'user' })
                    .withRecipient({ id: 'bot-id', name: 'Bot', role: 'bot' }, true)
                    .withChannelId('test-channel')
                    .withConversation({ id: 'test-conversation', conversationType: 'channel', isGroup: false })
                    .withId('1772129782775');
                context = buildActivityContext(activity);
                await context.reply('Here is your agenda');
                expect(mockSender.send).toHaveBeenCalledTimes(1);
                expect(mockSender.send).toHaveBeenCalledWith(expect.objectContaining({
                    type: 'message',
                    entities: expect.arrayContaining([
                        expect.objectContaining({
                            type: 'targetedMessageInfo',
                            messageId: '1772129782775',
                        }),
                    ]),
                }), mockRef);
            });
        });
    });
    describe('signin/signout flow', () => {
        beforeEach(() => {
            context = buildActivityContext(buildIncomingMessageActivity('Test message'));
        });
        it('returns existing token if available', async () => {
            const expectedToken = 'test-token';
            mockApiClient.users.token.get.mockResolvedValueOnce({
                token: expectedToken,
                connectionName: 'test-connection',
                channelId: 'test-channel',
                expiration: new Date(Date.now() + 1000 * 60 * 60).getTime().toString(),
            });
            const result = await context.signin();
            expect(result).toBe(expectedToken);
            expect(mockApiClient.users.token.get).toHaveBeenCalledWith({
                channelId: 'test-channel',
                userId: 'test-user',
                connectionName: 'test-connection',
            });
        });
        it('creates oauth card for new signin in 1:1 chat', async () => {
            mockApiClient.users.token.get.mockRejectedValueOnce(new Error('No token'));
            const mockResource = {
                tokenExchangeResource: {
                    uri: 'my-token-exhcange-resource-uri',
                },
                tokenPostResource: {
                    sasUrl: 'my-token-post-resource-sas-url',
                },
                signInLink: 'https://login.url',
            };
            mockApiClient.bots.signIn.getResource.mockResolvedValueOnce(mockResource);
            await context.signin();
            expect(mockSender.send).toHaveBeenCalledWith(expect.objectContaining({
                type: 'message',
                attachments: [
                    {
                        content: {
                            buttons: [
                                {
                                    title: 'Sign In',
                                    type: 'signin',
                                    value: 'https://login.url',
                                },
                            ],
                            connectionName: 'test-connection',
                            text: 'Please Sign In...',
                            tokenExchangeResource: {
                                uri: 'my-token-exhcange-resource-uri',
                            },
                            tokenPostResource: { sasUrl: 'my-token-post-resource-sas-url' },
                        },
                        contentType: 'application/vnd.microsoft.card.oauth',
                    },
                ],
                inputHint: 'acceptingInput',
                recipient: { id: 'test-user', name: 'Test User', role: 'user' },
            }), mockRef);
        });
        it('creates new 1:1 conversation for group chat signin', async () => {
            context = new activity_1.ActivityContext({
                ...context,
                activity: {
                    ...buildIncomingMessageActivity('Test message'),
                    conversation: {
                        id: 'group-conv',
                        isGroup: true,
                        conversationType: 'group',
                    },
                },
                activitySender: mockSender,
            });
            mockApiClient.users.token.get.mockRejectedValueOnce(new Error('No token'));
            mockApiClient.conversations.create.mockResolvedValueOnce({
                id: 'new-conv-id',
                activityId: 'new-activity-id',
                serviceUrl: '',
            });
            const mockResource = {
                tokenExchangeResource: {},
                tokenPostResource: {},
                signInLink: 'https://login.url',
            };
            mockApiClient.bots.signIn.getResource.mockResolvedValueOnce(mockResource);
            await context.signin();
            expect(mockApiClient.conversations.create).toHaveBeenCalled();
            expect(mockSender.send).toHaveBeenCalledWith(expect.objectContaining({ type: 'message', text: 'Please Sign In...' }), expect.any(Object));
        });
        it('forwards signout request to api client', async () => {
            await context.signout();
            expect(mockApiClient.users.token.signOut).toHaveBeenCalledWith({
                channelId: 'test-channel',
                userId: 'test-user',
                connectionName: 'test-connection',
            });
        });
    });
    describe('constructor — prototype method shadowing', () => {
        it('drops context properties that would shadow prototype methods', () => {
            const activity = buildIncomingMessageActivity('Hello world');
            const malicious = {
                send: jest.fn(),
                reply: jest.fn(),
                quote: jest.fn(),
                signin: jest.fn(),
                signout: jest.fn(),
            };
            const ctx = new activity_1.ActivityContext({
                appId: 'test-app',
                activity,
                ref: mockRef,
                log: mockLogger,
                api: mockApiClient,
                appGraph: {},
                userGraph: {},
                storage: mockStorage,
                connectionName: 'test-connection',
                next: jest.fn(),
                activitySender: mockSender,
                ...malicious,
            });
            for (const name of ['send', 'reply', 'quote', 'signin', 'signout']) {
                expect(Object.prototype.hasOwnProperty.call(ctx, name)).toBe(false);
                expect(ctx[name]).toBe(activity_1.ActivityContext.prototype[name]);
            }
        });
        it('still allows new properties from extra context', () => {
            const activity = buildIncomingMessageActivity('Hello world');
            const ctx = new activity_1.ActivityContext({
                appId: 'test-app',
                activity,
                ref: mockRef,
                log: mockLogger,
                api: mockApiClient,
                appGraph: {},
                userGraph: {},
                storage: mockStorage,
                connectionName: 'test-connection',
                next: jest.fn(),
                activitySender: mockSender,
                customField: 'still here',
            });
            expect(ctx.customField).toBe('still here');
        });
        it('routes ctx.send() to the prototype method even when a colliding key is supplied', async () => {
            const activity = buildIncomingMessageActivity('Hello world');
            const maliciousSend = jest.fn();
            const ctx = new activity_1.ActivityContext({
                appId: 'test-app',
                activity,
                ref: mockRef,
                log: mockLogger,
                api: mockApiClient,
                appGraph: {},
                userGraph: {},
                storage: mockStorage,
                connectionName: 'test-connection',
                next: jest.fn(),
                activitySender: mockSender,
                // Simulates a plugin's onActivity context attempting to inject its own send.
                send: maliciousSend,
            });
            await ctx.send({ type: 'message', text: 'real send' });
            expect(maliciousSend).not.toHaveBeenCalled();
            expect(mockSender.send).toHaveBeenCalledTimes(1);
        });
    });
});
